### PR TITLE
Fix 159

### LINKED ISSUE #159: Update question 307
### Thanks for helping!

Thank you for helping to port [this question](../blob/master/questions/307/) from C++17 to C++23.

You'll find the source code in [question.cpp](../blob/master/questions/307/question.cpp), this should normally not need modification.

In C++17, the correct answer is:
> The program is guaranteed to output: `0`

Please verify that this is still correct in C++23. If it is, please do the following:
- [ ] Update [explanation.md](../blob/master/questions/307/explanation.md):
  - [ ] Refer to the correct section numbers
  - [ ] Use updated quotes from those sections (the wording might have changed)
  - [ ] Make sure the rest of the text in the explanation is consistent with the new standard
- [ ] Update [hint.md](../blob/master/questions/307/hint.md) if needed (usually not needed)

*Note that some older questions are missing references to the standard, we don't need to fix that now.*

If you get stuck, please describe the problem in detail in a comment to the issue, and a maintainer will add the `help wanted` tag so others can chime in.

If the correct answer has changed from C++17 to C++23, you can either just leave a comment in this issue, or see the instructions for [updating meta data](../blob/master/METADATA_HOWTO.md).

### PR DESCRIPTION
**Pull request by knatten**
Fixes #159 

This replaces PR #192. I re-ordered the commits so that @eoan-ermine has the last commit on the branch and counts as the person who fixed it.

### CODE REVIEW

File: questions/307/explanation.md [hunk 1]
```diff
@@ -1,21 +1,13 @@
...
\ No newline at end of file
+According to the [dcl.fct.def.delete]¶2:
+
+> A program that refers to a deleted function implicitly or explicitly, other than to declare it, is ill-formed.
+
+Does `S{}` refer to the deleted constructor? That depends on whether this is value-initialization or aggregate initialization. In the case of value-initialization, the constructor will be called, but in the case of aggregate initialization, no constructor is called.
+
+The rules for whether a struct is an aggregate are quite complex §[dcl.init.aggr]¶1:
+
+> An aggregate is an array or a class with no user-declared or inherited constructors, no private or protected non-static data members, no virtual functions, and no virtual, private, or protected base classes.
```
Comment (tocic) [inline questions/307/explanation.md:9]: nit: I would quote this the same as in the standard (i.e., line-by-line enumeration) or provide the reference after this sentence in parentheses

File: questions/307/meta_data.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
     "id": 307,
-    "result": "OK",
-    "answer": "0",
+    "result": "CE",
+    "answer": "",
     "state": "PUB",
     "difficulty": 2
-}
\ No newline at end of file
+}
```
Comment (tocic) [inline questions/307/meta_data.json:7]: nit: usually, there's no EOL in `meta_data.json`

File: questions/307/explanation.md [hunk 3]
```diff
@@ -1,21 +1,13 @@
...
+
+Here, the `S` default constructor is considered user-declared, so `S` isn't an aggregate, and `S{}` is using value initialization, and the program is ill-formed.
+
+Note that this behavior changed in C++20 (P1008R1). Previously, a constructor that was only user-declared but not user-provided did not prevent a class from being an aggregate. So prior to C++20, this question did actually compile.
```
Comment (tocic) [inline questions/307/explanation.md:13]:
```suggestion
Note that this behavior changed in C++20 ([P1008](https://wg21.link/p1008)). Previously, a constructor that was only user-declared but not user-provided did not prevent a class from being an aggregate. So prior to C++20, this question did actually compile.
```

**Review by tocic — APPROVED** [review]

File: questions/307/meta_data.json (hunk 2 above)
Comment (knatten) [reply questions/307/meta_data.json:7]: This doesn't matter, these files are just read by a script and die with this repo.

**Pull request merged**